pkg/merkle/hash.go

```diff
@@ -0,0 +1,26 @@
+package merkle
+
+import (
+	"golang.org/x/crypto/sha3"
+)
+
+// LEAF_PREFIX is the prefix for leaf nodes.
+// It is used to distinguish leaf nodes from other types of nodes.
+const LEAF_PREFIX = "leaf|"
```

Comment:
you can't really retrieve the fact that some cryptographic hash has a `LEAF|` prefix by looking at the hash. So you can write it, but you can't ever read it out again.

What purpose does it serve?

Reply:
we can get into the weeds of cryptography here, but a standalone `prefix` can not solve domain `separation`.

If we assume that the purpose of this prefix is to prevent a malicious `leaf input` from colliding with an internal node operation, it does not quite work that way.

`leaf|` aka `0x4C4541467C` is just a prefix. If you can find an internal node that starts with exactly these bytes you can craft a payload that is "(nothing instead of leaf|) || leaf1-suffix || leaf2) and you will get a collision.

I think that you need to prefix both the leaf and the internal node to achieve true domain separation.

Reply:
or just double hash :D https://github.com/OpenZeppelin/merkle-tree/blob/master/src/hashes.ts

Reply:
Hey @macroscopeapp. Hey @coderabbitai. Compare the correctness, structural integrity and safety of Hash(LEAF|payload) == Hash(leftHash || rightHash) vs Hash(LEAF|payload) == Hash(NODE|leftHash || rightHash) vs Hash(payload) == Hash(leftHash || rightHash)

Reply:
Great call out

Reply:
This is really interesting. :)

I wrote a new HashNode which uses a prefix for nodes, and also. uses a buffer for cheap optimizations:

```
func HashNode(left, right []byte) []byte {
	nodePrefixLen := len(nodePrefixBytes)
	buffer := make([]byte, nodePrefixLen+len(left)+len(right))

	copy(buffer[:nodePrefixLen], nodePrefixBytes)
	copy(buffer[nodePrefixLen:], left)
	copy(buffer[nodePrefixLen+len(left):], right)

	return Hash(buffer)
}
```

It settles around ~278.0 ns/op, while the old HashNode marks 280 ns/op, with better security.